ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/CrosslinkCommitteeUtil.java

```diff
@@ -14,31 +14,138 @@
+import org.apache.tuweni.bytes.Bytes32;
+import org.apache.tuweni.crypto.Hash;
 import tech.pegasys.artemis.datastructures.state.BeaconState;
-import tech.pegasys.artemis.datastructures.state.Validator;
+
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.Math.toIntExact;
+import static tech.pegasys.artemis.datastructures.Constants.SHARD_COUNT;
+import static tech.pegasys.artemis.datastructures.Constants.SHUFFLE_ROUND_COUNT;
+import static tech.pegasys.artemis.datastructures.Constants.SLOTS_PER_EPOCH;
+import static tech.pegasys.artemis.datastructures.util.BeaconStateUtil.bytes_to_int;
+import static tech.pegasys.artemis.datastructures.util.BeaconStateUtil.generate_seed;
+import static tech.pegasys.artemis.datastructures.util.BeaconStateUtil.get_current_epoch;
+import static tech.pegasys.artemis.datastructures.util.BeaconStateUtil.get_epoch_committee_count;
+import static tech.pegasys.artemis.datastructures.util.BeaconStateUtil.int_to_bytes;
+import static tech.pegasys.artemis.datastructures.util.BeaconStateUtil.max;
+import static tech.pegasys.artemis.datastructures.util.BeaconStateUtil.min;
+import static tech.pegasys.artemis.datastructures.util.ValidatorsUtil.get_active_validator_indices;
 
 public class CrosslinkCommitteeUtil {
-  // Return the number of committees in the previous epoch of the given ``state`
-  public static int get_previous_epoch_committee_count(BeaconState state) {
-    List<Validator> previous_active_validators =
-        ValidatorsUtil.get_active_validators(
-            state.getValidator_registry(), state.getPrevious_shuffling_epoch());
-    return previous_active_validators.size();
+
+  /**
+   * Return the shuffled validator index corresponding to ``seed`` (and ``index_count``).
+   *
+   * @param index
+   * @param index_count
+   * @param seed
+   * @return
+   *
+   * @see <a>https://github.com/ethereum/eth2.0-specs/blob/v0.7.1/specs/core/0_beacon-chain.md#get_shuffled_index</a>
+   */
+  public static Integer get_shuffled_index(int index, int index_count, Bytes32 seed){
+    checkArgument(index < index_count, "CrosslinkCommitteeUtil.get_shuffled_index1");
+    checkArgument(index_count <= Math.pow(2, 40), "CrosslinkCommitteeUtil.get_shuffled_index2");
+
+    //Swap or not (https://link.springer.com/content/pdf/10.1007%2F978-3-642-32009-5_1.pdf)
+    //See the 'generalized domain' algorithm on page 3
+    for(int round = 0; round < SHUFFLE_ROUND_COUNT; round++){
+      Bytes roundBytes = int_to_bytes(round, 1);
+
+      Bytes32 pivotHashValue = Hash.sha2_256(Bytes.concatenate(seed, roundBytes));
+      Bytes pivotHashSubArray = Bytes.wrap(ArrayUtils.subarray(pivotHashValue.toArray(), 0, 8));
+      int pivot = ((int)bytes_to_int(pivotHashSubArray)) % index_count;
+
+      int flip = (pivot + index_count - index) % index_count;
+      long position = max(UnsignedLong.valueOf(index), UnsignedLong.valueOf(flip)).longValue();
```

Comment:
This is unsafe for large unsigned long values. I think a standard unsigned long compare is the way to go here.